src/iocp/afd.rs

```diff
@@ -0,0 +1,608 @@
+        in_use: AtomicBool,
+
+        // The auxillary data.
+        #[pin]
+        data: T,
+
+        // This block is not allowed to move.
+        #[pin]
+        _marker: PhantomPinned,
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for IoStatusBlock<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("IoStatusBlock")
+            .field("iosb", &"..")
+            .field("in_use", &self.in_use)
+            .field("data", &self.data)
+            .finish()
+    }
+}
+
+impl<T> From<T> for IoStatusBlock<T> {
+    fn from(data: T) -> Self {
+        Self {
+            iosb: UnsafeCell::new(unsafe { std::mem::zeroed() }),
+            in_use: AtomicBool::new(false),
+            data,
+            _marker: PhantomPinned,
+        }
+    }
+}
+
+impl<T> IoStatusBlock<T> {
+    pub(super) fn iosb(self: Pin<&Self>) -> &UnsafeCell<IO_STATUS_BLOCK> {
+        self.project_ref().iosb
+    }
+
+    pub(super) fn data(self: Pin<&Self>) -> Pin<&T> {
+        self.project_ref().data
+    }
+}
+
+impl<T: HasAfdInfo> HasAfdInfo for IoStatusBlock<T> {
+    fn afd_info(self: Pin<&Self>) -> Pin<&UnsafeCell<AfdPollInfo>> {
+        self.project_ref().data.afd_info()
+    }
+}
+
+/// Can be transmuted to an I/O status block.
+///
+/// # Safety
+///
+/// A pointer to `T` must be able to be converted to a pointer to `IO_STATUS_BLOCK`
+/// without any issues.
+pub(super) unsafe trait AsIoStatusBlock {}
+
+unsafe impl<T> AsIoStatusBlock for IoStatusBlock<T> {}
+unsafe impl<T> Completion for IoStatusBlock<T> {
+    fn try_lock(self: Pin<&Self>) -> bool {
+        !self.in_use.swap(true, Ordering::SeqCst)
+    }
+
+    unsafe fn unlock(self: Pin<&Self>) {
+        self.in_use.store(false, Ordering::SeqCst);
+    }
+}
+
+/// Get the base socket associated with a socket.
+pub(super) fn base_socket(sock: RawSocket) -> io::Result<RawSocket> {
+    // First, try the SIO_BASE_HANDLE ioctl.
+    let result = unsafe { try_socket_ioctl(sock, SIO_BASE_HANDLE) };
+
+    match result {
+        Ok(sock) => return Ok(sock),
+        Err(e) if e.kind() == io::ErrorKind::InvalidInput => return Err(e),
+        Err(_) => {}
+    }
+
+    // Some poorly coded LSPs may not handle SIO_BASE_HANDLE properly, but in some cases may
+    // handle SIO_BSP_HANDLE_POLL better. Try that.
+    let result = unsafe { try_socket_ioctl(sock, SIO_BSP_HANDLE_POLL)? };
+    if result == sock {
+        return Err(io::Error::from(io::ErrorKind::InvalidInput));
+    }
+
+    // Try `SIO_BASE_HANDLE` again, in case the LSP fixed itself.
+    unsafe { try_socket_ioctl(result, SIO_BASE_HANDLE) }
+}
+
+/// Run an IOCTL on a socket and return a socket.
+///
+/// # Safety
+///
+/// The `ioctl` parameter must be a valid I/O control that returns a valid socket.
+unsafe fn try_socket_ioctl(sock: RawSocket, ioctl: u32) -> io::Result<RawSocket> {
+    let mut out = MaybeUninit::<RawSocket>::uninit();
+    let mut bytes = MaybeUninit::<u32>::uninit();
```

Comment:
what do we gain by making this `u32` `MaybeUninit` here?

Reply:
Nothing, I guess. Fixed.